### DISCUSSION TITLE
Make it easier to view change in diff, then accept it or not.

### DISCUSSION THREAD

**Discussion by bassamsdata** [olimorris/codecompanion.nvim#154] [Ideas]
Hello,

I’d like to suggest a feature similar to avante.nvim. In chat, it could display the lines that would be changed, and by pressing `A` with the cursor in the chat code block, it would add the changes to the buffer. From there, the diff would show both versions, allowing me to press `ct` to choose the new version or `co` to keep mine, and so on. 

To summarize, the typical process would be: press `A`, then `co` or `ct`, and that’s it.

Why I’m suggesting this:
I’m fully aware of the inline options (and thank you for that as well), but asking the LLM to make changes after a chat can be a bit slow, to be honest. Pressing a single key like `A` is much more efficient. Additionally, it opens multiple buffers in diff mode, and when I want to revert changes, I have to press `u` several times, which is much slower than simply pressing `co`. I know someone already suggested something similar for diff in [#109](#109), but this is a bit different.

Thank you so much.

![Kapture 2024-08-31 at 00 37 24](https://github.com/user-attachments/assets/910fbb60-fc71-4533-9b37-076aec8acac9)

**Comment by bassamsdata** [comment]
I just noticed the `gr` and `ga` keymappings that work within the diff—my mistake for overlooking them. 

However, I still believe that having the diff in a single buffer is much easier, though I understand this might be a personal preference. I also recognize the value of relying on Neovim's internal diff mode. 
I’ll likely explore implementing something using the `Mini.diff` module. That said, I still find the idea of adding code blocks directly from chat to be quite useful.

Thank you.

**Reply by olimorris** [reply]
Hey @bassamsdata.

A long term aim is to be able to update a buffer from the chat buffer. That will be in the plugin at some point in the near future. This week I'm going to focus on implementing slash commands from Zed AI. That will open up a huge amount of productivity for devs.

Btw, I added inline transformations to the plugin earlier in the week. So the inline assistant has knowledge of the code that the LLM has generated in a chat buffer #130.

**Reply by olimorris** [reply]
And I would be totally up for `Mini.diff` support being added. Without looking at the plugin, I imagine I already have a lot of the requirements built into the `inline.lua` file.

**Comment by bassamsdata** [comment]
Thank you for your response.

> That will be in the plugin at some point in the near future.

This is great news.

> implementing slash commands from Zed AI

This would be very useful indeed, I played a little with Zed's `/prompts` and that was fun.

> inline assistant has knowledge of the code that the LLM has generated in a chat buffer.

I've tested it, but for some reason, I always find myself reverting back to chat copy-and-paste since I use refactor only most of the time.

> I imagine I already have a lot of the requirements built into the inline.lua file.

I'll play with it a bit to see if I can set something up and get it working.

Thank you again.

**Reply by olimorris** [reply]
> I've tested it, but for some reason, I always find myself reverting back to chat copy-and-paste since I use refactor only most of the time.

Yeah this is a little bit cumbersome right now. You can visually select the text and it should overwrite it but appreciate that's about as efficient as copying and pasting yourself.

**Comment by bassamsdata** [comment]
Hey @olimorris,

I’ve created a [module](https://github.com/bassamsdata/llmdiff.nvim) that utilizes `mini.diff` with codecompanion for the inline feature, showing the diff directly in the buffer.

Here’s a quick demo.
![Kapture 2024-09-10 at 00 33 31](https://github.com/user-attachments/assets/ae59defb-353a-4ac9-8317-a0b0c7a38b1d)


I can submit a PR to integrate this into `codecompanion` so users can easily opt-in as an alternative to the original nvim-diff, or I can keep it as an extension for the plugin. What do you think?

Thank you!

**Reply by olimorris** [reply]
Oh that would be amazing! I had this on my todo list to explore this week but I would love to see this as a PR. We can workout how best to integrate this but I think it's an awesome alternative.

**Reply by raine** [reply]
FWIW, I would welcome this proposed way of showing the diff!

**Reply by raine** [reply]
Hmm, noticed this is added already. So I installed [echasnovski/mini.diff](https://github.com/echasnovski/mini.diff) and changed diff provider to `mini_diff`. However, I'm not seeing any diff as result. The plugin replaces the code according to instructions and it can be reverted with `gr`, but it's not a diff visually, just the resulting code.

**Reply by olimorris** [reply]
Is your config correct? I moved it around yesterday. The diff should sit within the `display` table.

I was using it this morning and it seems to be working fine.

**Reply by bassamsdata** [reply]
@raine if you still have an issue, Would you mind sharing your mini.diff config along with your codecompanion config?

**Reply by raine** [reply]
@bassamsdata Thanks, I have no config for `mini.diff`, should I? 😄 

```lua
require("codecompanion").setup({
	display = {
		diff = {
			provider = "mini_diff",
		},
	},
	strategies = {
		chat = {
			adapter = "anthropic",
		},
		inline = {
			adapter = "anthropic",
		},
		agent = {
			adapter = "anthropic",
		},
	},
})

```

**Reply by bassamsdata** [reply]
@raine Well, I think it's `lazy = true` by default in lazy.nvim plugin manager, so codecompanion probably won't detect it. I suggest calling the setup function in mini.diff as well, or at least `opts = {}` as `dependencies` for codecompanion.

```lua
    dependencies = {
      { "echasnovski/mini.diff", opts = {} },
    },
```

I have mine set to lazy load on buffer for these events so it won't affect any nvim startup speed.
```lua
{
    "echasnovski/mini.diff",
event = { "BufReadPost", "BufNewFile", "BufWritePre" }
}
```

**Reply by olimorris** [reply]
I disable it by default but have it setup with:

```lua
  {
    "echasnovski/mini.diff",
    config = function()
      local diff = require("mini.diff")
      diff.setup({
        -- Disabled by default
        source = diff.gen_source.none(),
      })
    end,
  },
```

**Reply by jacobdaddario** [reply]
I don't know why, but I couldn't for the life of me get the `mini_diff` provider to work. I tried setting `lazy = false` among other things. @olimorris 's suggestion is what made it work.

Would love to know why it fixed things.

**Comment by xheisenbugx** [comment]
Hi, is it possible to apply individual hunks using `mini.diff`?

I'm using the following configuration:

```
display = {
        diff = {
          enabled = true,
          close_chat_at = 240, -- Close an open chat buffer if the total columns of your display are less than...
          layout = "vertical", -- vertical|horizontal split for default provider
          opts = { "internal", "filler", "closeoff", "algorithm:patience", "followwrap", "linematch:120" },
          provider = "mini_diff", -- default|mini_diff
        },
        chat = {
          -- Change to true to show the current model
          show_settings = false,
          window = {
            layout = "vertical", -- float|vertical|horizontal|buffer
          },
        },
      }
```

After performing a refactor, it generates a diff buffer, but when I try to apply a specific hunk using `vip` followed by `gh`, I receive this error:

```
E5108: Error executing lua (mini.diff) Current source does not support applying hunks.
stack traceback:
	[C]: in function 'error'
	...valdo/.local/share/nvim/lazy/mini.diff/lua/mini/diff.lua:1759: in function 'error'
	...valdo/.local/share/nvim/lazy/mini.diff/lua/mini/diff.lua:1129: in function 'apply_hunks'
	...valdo/.local/share/nvim/lazy/mini.diff/lua/mini/diff.lua:739: in function 'do_hunks'
	...valdo/.local/share/nvim/lazy/mini.diff/lua/mini/diff.lua:821: in function <...valdo/.local/share/nvim/lazy/mini.diff/lua/mini/diff.lua:803>
```

It seems that I can only apply all hunks at once using `ga` or reject them with `gr`. Is there a way to apply individual hunks?

Please let me know if I am doing something wrong, I am using the mini.diff plugin provided by LazyExtras.

Also, I don't have a lot of experiences with diff plugins. I would appreciate your help, thanks in advance!

**Reply by olimorris** [reply]
Hi. It's not possible. I'm only adding very basic support for diffs right now owing to the complexity of maintaining so many features in the plugin.

I suspect the issue you're encountering is because we create a new source for CodeCompanion changes and only the git source would allow for hunks.

If you want to investigate the issue then I'll be open to a PR but we have a strict diff structure for both `mini.diff` and the `default` implementation.

**Comment by carlosflorencio** [comment]
I would also appreciate the style of diff `avante` uses (git conflict inline). Keep up the good work 💪.
